src/Common/src/System/Collections/Generic/EnumerableHelpers.Linq.cs

```diff
@@ -0,0 +1,46 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+
+namespace System.Collections.Generic
+{
+    /// <summary>Internal helper functions for working with enumerables.</summary>
+    internal static partial class EnumerableHelpers
+    {
+        /// <summary>Gets the count of the specified enumerable.</summary>
+        /// <param name="source">The enumerable to count.</param>
+        /// <returns>The count of the enumerable.</returns>
+        internal static int Count<T>(IEnumerable<T> source)
+        {
+            Debug.Assert(source != null);
+
+            var collectionOfT = source as ICollection<T>;
+            if (collectionOfT != null)
+            {
+                return collectionOfT.Count;
+            }
+            
+            var collection = source as ICollection;
+            if (collection != null)
+            {
+                return collection.Count;
+            }
+            
+            int count = 0;
+            using (IEnumerator<T> e = source.GetEnumerator())
```

Comment:
> But since we are running the selector anyway

There isn't a selector in cases like:
https://github.com/dotnet/corefx/pull/12703/files#diff-e6e91d17f21cf11b8d7b5ba1c23c933aR110

> I don't think it's worth trying to optimize these functions

Why?

> we could have just written the whole thing inline like I did here

Yes, like I suggested at https://github.com/dotnet/corefx/pull/12703/files#r87900623, which got a thumbs up from you but doesn't appear to have been addressed, so I'm unclear what the plan is.